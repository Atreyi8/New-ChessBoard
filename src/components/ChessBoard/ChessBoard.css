.chessboard {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 20px auto;
    border: 2px solid black;
    width: fit-content;
  }
  
  .row {
    display: flex;
  }
  
  
  