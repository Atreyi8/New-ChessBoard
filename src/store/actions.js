
export const ActionTypes = {
    TILE_CLICK: "tile_click",
  };
  

  export const tileClick = (row, col) => ({
    type: ActionTypes.TILE_CLICK,
    payload: { row, col },
  });
  


  