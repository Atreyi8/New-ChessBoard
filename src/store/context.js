import {createContext , useContext} from "react"
export const ChessBoardContext = createContext()

export function useChessBoardContext(){
  return useContext(ChessBoardContext)
}