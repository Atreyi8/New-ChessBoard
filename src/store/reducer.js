export const initialBoardState = [
    [
        { type: "rook", color: "black", symbol: "♜" },
        { type: "knight", color: "black", symbol: "♞" },
        { type: "bishop", color: "black", symbol: "♝" },
        { type: "queen", color: "black", symbol: "♛" },
        { type: "king", color: "black", symbol: "♚" },
        { type: "bishop", color: "black", symbol: "♝" },
        { type: "knight", color: "black", symbol: "♞" },
        { type: "rook", color: "black", symbol: "♜" },
    ],
    Array(8).fill({ type: "pawn", color: "black", symbol: "♟" }),
    ...Array(4).fill(Array(8).fill(null)),
    Array(8).fill({ type: "pawn", color: "white", symbol: "♙" }),
    [
        { type: "rook", color: "white", symbol: "♖" },
        { type: "knight", color: "white", symbol: "♘" },
        { type: "bishop", color: "white", symbol: "♗" },
        { type: "queen", color: "white", symbol: "♕" },
        { type: "king", color: "white", symbol: "♔" },
        { type: "bishop", color: "white", symbol: "♗" },
        { type: "knight", color: "white", symbol: "♘" },
        { type: "rook", color: "white", symbol: "♖" },
    ],
];

export const createInitialState = () => {
    return {
        board: initialBoardState,
        selectedTile: []
    }
}
export function reducer(state, action) {
  switch (action.type) {
    case "tile_click": {
      const { row, col } = action.payload;
      const tileIndex = state.selectedTile.findIndex(
        (tile) => tile.row === row && tile.col === col
      );

      // If tile is already selected remove it , or add it
      const updatedSelectedTile =
        tileIndex >= 0
          ? state.selectedTile.filter((_, index) => index !== tileIndex) // remove
          : [...state.selectedTile, { row, col }]; //add

      return {
        ...state,
        selectedTile: updatedSelectedTile,
      };
    }
    default:
      return state;
  }
}

