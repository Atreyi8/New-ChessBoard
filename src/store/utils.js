const isValidMove = ({ row, col }) => row >= 0 && row < 8 && col >= 0 && col < 8;

// Diagonal moves (Bishop and Queen)
const getDiagonalMoves = (row, col) => {
  const moves = [];
  for (let i = 1; i < 8; i++) {
    moves.push({ row: row + i, col: col + i }); // Bottom-right
    moves.push({ row: row + i, col: col - i }); // Bottom-left
    moves.push({ row: row - i, col: col + i }); // Top-right
    moves.push({ row: row - i, col: col - i }); // Top-left
  }
  return moves.filter(isValidMove);
};

// Horizontal and vertical moves (Rook and Queen)
const getStraightLineMoves = (row, col) => {
  const moves = [];
  for (let i = 1; i < 8; i++) {
    moves.push({ row: row + i, col }); // Down
    moves.push({ row: row - i, col }); // Up
    moves.push({ row, col: col + i }); // Right
    moves.push({ row, col: col - i }); // Left
  }
  return moves.filter(isValidMove);
};

// L shaped moves (Knight)
const getKnightMoves = (row, col) => {
  const moves = [
    { row: row + 2, col: col + 1 },
    { row: row + 2, col: col - 1 },
    { row: row - 2, col: col + 1 },
    { row: row - 2, col: col - 1 },
    { row: row + 1, col: col + 2 },
    { row: row + 1, col: col - 2 },
    { row: row - 1, col: col + 2 },
    { row: row - 1, col: col - 2 },
  ];
  return moves.filter(isValidMove);
};

// One step moves (King)
const getOneStepMoves = (row, col) => {
  const moves = [
    { row: row + 1, col },
    { row: row - 1, col },
    { row, col: col + 1 },
    { row, col: col - 1 },
    { row: row + 1, col: col + 1 },
    { row: row + 1, col: col - 1 },
    { row: row - 1, col: col + 1 },
    { row: row - 1, col: col - 1 },
  ];
  return moves.filter(isValidMove);
};

// Pawn moves (with direction)
const getPawnMoves = (row, col, tile) => {
  const moves = [];
  const direction = tile.color === "white" ? -1 : 1; // Up for white, down for black
  const forwardMove = { row: row + direction, col };
  if (isValidMove(forwardMove)) {
    moves.push(forwardMove);
  }
  return moves;
};
export const calculateValidMoves = (row, col, tile) => {
    let validMoves = [];
  
    switch (tile.type) {
      case "pawn":
        validMoves = getPawnMoves(row, col, tile);
        break;
  
      case "knight":
        validMoves = getKnightMoves(row, col);
        break;
  
      case "rook":
        validMoves = getStraightLineMoves(row, col);
        break;
  
      case "bishop":
        validMoves = getDiagonalMoves(row, col);
        break;
  
      case "queen":
        validMoves = [...getDiagonalMoves(row, col), ...getStraightLineMoves(row, col)];
        break;
  
      case "king":
        validMoves = getOneStepMoves(row, col);
        break;
  
      default:
        break;
    }
  
    return validMoves;
  };
  
  export const initialBoardState = [
    [
        { type: "rook", color: "black", symbol: "♜" },
        { type: "knight", color: "black", symbol: "♞" },
        { type: "bishop", color: "black", symbol: "♝" },
        { type: "queen", color: "black", symbol: "♛" },
        { type: "king", color: "black", symbol: "♚" },
        { type: "bishop", color: "black", symbol: "♝" },
        { type: "knight", color: "black", symbol: "♞" },
        { type: "rook", color: "black", symbol: "♜" },
    ],
     Array(8).fill({ type: "pawn", color: "black", symbol: "♟" }),
    // [ { type: "rook", color: "black", symbol: "♜" },
    //   { type: "knight", color: "black", symbol: "♞" },
    //   { type: "bishop", color: "black", symbol: "♝" },
    //   { type: "queen", color: "black", symbol: "♛" },
    //   { type: "king", color: "black", symbol: "♚" },
    //   { type: "bishop", color: "black", symbol: "♝" },
    //   { type: "knight", color: "black", symbol: "♞" },
    //   { type: "rook", color: "black", symbol: "♜" },],
    // Array(4).fill({ type: "knight", color: "black", symbol: "♞" }),
    // Array(4).fill({ type: "rook", color: "black", symbol: "♜" }),
    ...Array(4).fill(Array(8).fill(null)),
    Array(8).fill({ type: "pawn", color: "white", symbol: "♙" }),
    [
        { type: "rook", color: "white", symbol: "♖" },
        { type: "knight", color: "white", symbol: "♘" },
        { type: "bishop", color: "white", symbol: "♗" },
        { type: "queen", color: "white", symbol: "♕" },
        { type: "king", color: "white", symbol: "♔" },
        { type: "bishop", color: "white", symbol: "♗" },
        { type: "knight", color: "white", symbol: "♘" },
        { type: "rook", color: "white", symbol: "♖" },
    ],
];

