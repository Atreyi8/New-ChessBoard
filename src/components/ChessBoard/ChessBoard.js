import React from "react";
import Tile from "../Tile/Tile";
import "./ChessBoard.css"
import { useChessBoardContext } from "../../store/context";

const ChessBoard = () => {
  const { state, dispatch } = useChessBoardContext();
  const { board, selectedTile } = state; 

  const handleTileClick = (row, col) => {
    console.log(`Tile clicked at row: ${row}, col: ${col}`);
    console.log(row,col)
    dispatch({ type: "tile_click", payload: { row, col } });
  };

  return (
    <div className="chessboard">
      {console.log(state)}
      {board.map((row, rowIndex) => (
        <div key={rowIndex} className="row">
          {row.map((tile, colIndex) => (
            <Tile
              key={`${rowIndex}-${colIndex}`}
              row={rowIndex}
              col={colIndex}
              tileData={tile}
              onClick={() => handleTileClick(rowIndex, colIndex)}
              isSelected={selectedTile.some(
                (tile) => tile.row === rowIndex && tile.col === colIndex
              )}
            />
          ))}
        </div>
      ))}
    </div>
  );
};

export default ChessBoard;
