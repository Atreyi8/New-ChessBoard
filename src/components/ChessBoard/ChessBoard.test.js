import { render, screen, fireEvent } from "@testing-library/react";
import ChessBoard from "./ChessBoard";
import { ChessBoardContext } from "../../store/context"; // Mocking the context provider

// Mock Tile component
jest.mock("../Tile/Tile", () => ({ row, col, onClick, isSelected }) => (
  <button
    data-testid={`tile-${row}-${col}`}
    className={isSelected ? "selected" : ""}
    onClick={onClick}
  >
    Tile {row},{col}
  </button>
));

describe("ChessBoard Component", () => {
  const mockDispatch = jest.fn();

  const mockContextValue = {
    state: {
      board: [
        ["", "", ""],
        ["", "", ""],
        ["", "", ""],
      ],
      selectedTile: [{ row: 1, col: 1 }],
    },
    dispatch: mockDispatch,
  };

  const renderWithContext = (component) =>
    render(
      <ChessBoardContext.Provider value={mockContextValue}>
        {component}
      </ChessBoardContext.Provider>
    );

  test("renders the chessboard correctly", () => {
    renderWithContext(<ChessBoard />);
    // Assert the number of tiles rendered
    const tiles = screen.getAllByRole("button");
    expect(tiles).toHaveLength(9); // 3x3 board in mock context

    // Check specific tile rendering
    expect(screen.getByTestId("tile-1-1")).toHaveClass("selected");
    expect(screen.getByTestId("tile-0-0")).not.toHaveClass("selected");
  });

  test("calls dispatch when a tile is clicked", () => {
    renderWithContext(<ChessBoard />);

    // Simulate a tile click
    const tile = screen.getByTestId("tile-0-0");
    fireEvent.click(tile);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "tile_click",
      payload: { row: 0, col: 0 },
    });
  });

  test("matches the snapshot", () => {
    const { asFragment } = renderWithContext(<ChessBoard />);
    expect(asFragment()).toMatchSnapshot();
  });
});
