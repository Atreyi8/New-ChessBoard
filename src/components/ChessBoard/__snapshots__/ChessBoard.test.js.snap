// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChessBoard Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="chessboard"
  >
    <div
      class="row"
    >
      <button
        class=""
        data-testid="tile-0-0"
      >
        Tile 0,0
      </button>
      <button
        class=""
        data-testid="tile-0-1"
      >
        Tile 0,1
      </button>
      <button
        class=""
        data-testid="tile-0-2"
      >
        Tile 0,2
      </button>
    </div>
    <div
      class="row"
    >
      <button
        class=""
        data-testid="tile-1-0"
      >
        Tile 1,0
      </button>
      <button
        class="selected"
        data-testid="tile-1-1"
      >
        Tile 1,1
      </button>
      <button
        class=""
        data-testid="tile-1-2"
      >
        Tile 1,2
      </button>
    </div>
    <div
      class="row"
    >
      <button
        class=""
        data-testid="tile-2-0"
      >
        Tile 2,0
      </button>
      <button
        class=""
        data-testid="tile-2-1"
      >
        Tile 2,1
      </button>
      <button
        class=""
        data-testid="tile-2-2"
      >
        Tile 2,2
      </button>
    </div>
  </div>
</DocumentFragment>
`;
